x-pack/libbeat/management/managerV2.go

```diff
@@ -0,0 +1,300 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package management
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/gofrs/uuid"
+
+	"github.com/elastic/beats/v7/libbeat/common/reload"
+	lbmanagement "github.com/elastic/beats/v7/libbeat/management"
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// BeatV2Manager is the main type for tracing V2-related config updates
+type BeatV2Manager struct {
+	config   *Config
+	registry *reload.Registry
+	client   client.V2
+
+	logger *logp.Logger
+
+	// Track individual units given to us by the V2 API
+	unitsMut sync.Mutex
+	units    map[string]*client.Unit
+	mainUnit string
+
+	// This satisfies the SetPayload() function, and will pass along this value to the UpdateStatus()
+	// call whenever a config is re-registered
+	payload map[string]interface{}
+
+	// stop callback must be registered by libbeat, as with the V1 callback
+	stopFunc func()
```

Comment:
What is the threading model for the manager because it looks like more of the members need to be guarded by a mutex. Like the stopFunc can be set through an exported method and potentially read by another goroutine when stop is called.

Reply:
So, we should only ever spin up a single "main" thread for the controller, with separate threads every time we get a config update, since we don't want to block that channel from the client. Looking over the code I think you're right though, I don't particularly like how the shutdown/stopFunc is carted around, and I think I'm gonna refactor that.